duplex_http_call.go

```diff
@@ -131,160 +120,196 @@ func (d *duplexHTTPCall) CloseWrite() error {
+	if _, err := discard(c.response.Body); err != nil &&
 		!errors.Is(err, context.Canceled) &&
 		!errors.Is(err, context.DeadlineExceeded) {
-		_ = d.response.Body.Close()
+		_ = c.response.Body.Close()
 		return wrapIfRSTError(err)
 	}
-	return wrapIfRSTError(d.response.Body.Close())
+	return wrapIfRSTError(c.response.Body.Close())
 }
 
 // ResponseStatusCode is the response's HTTP status code.
-func (d *duplexHTTPCall) ResponseStatusCode() (int, error) {
-	if err := d.BlockUntilResponseReady(); err != nil {
+func (c *httpCall) ResponseStatusCode() (int, error) {
+	if err := c.BlockUntilResponseReady(); err != nil {
 		return 0, err
 	}
-	return d.response.StatusCode, nil
+	return c.response.StatusCode, nil
 }
 
 // ResponseHeader returns the response HTTP headers.
-func (d *duplexHTTPCall) ResponseHeader() http.Header {
-	_ = d.BlockUntilResponseReady()
-	if d.response != nil {
-		return d.response.Header
+func (c *httpCall) ResponseHeader() http.Header {
+	_ = c.BlockUntilResponseReady()
+	if c.response != nil {
+		return c.response.Header
 	}
 	return make(http.Header)
 }
 
 // ResponseTrailer returns the response HTTP trailers.
-func (d *duplexHTTPCall) ResponseTrailer() http.Header {
-	_ = d.BlockUntilResponseReady()
-	if d.response != nil {
-		return d.response.Trailer
+func (c *httpCall) ResponseTrailer() http.Header {
+	_ = c.BlockUntilResponseReady()
+	if c.response != nil {
+		return c.response.Trailer
 	}
 	return make(http.Header)
 }
 
 // SetValidateResponse sets the response validation function. The function runs
 // in a background goroutine.
-func (d *duplexHTTPCall) SetValidateResponse(validate func(*http.Response) *Error) {
-	d.validateResponse = validate
+func (c *httpCall) SetValidateResponse(validate func(*http.Response) *Error) {
+	c.validateResponse = validate
 }
 
 // BlockUntilResponseReady returns when the response is ready or reports an
 // error from initializing the request.
-func (d *duplexHTTPCall) BlockUntilResponseReady() error {
-	<-d.responseReady
-	return d.responseErr
+func (c *httpCall) BlockUntilResponseReady() error {
+	<-c.responseReady
+	return c.responseErr
+}
+
+func (c *httpCall) isClientStream() bool {
+	return c.streamType&StreamTypeClient != 0
+}
+
+func (c *httpCall) sendUnary(buffer *bytes.Buffer) error {
+	if !c.requestSent.CompareAndSwap(false, true) {
+		return fmt.Errorf("unary call already sent")
+	}
+	// Client request is unary, build the full request body and block on
+	// sending the request.
+	if c.request.Method != http.MethodGet && buffer != nil {
+		payload := newPayloadCloser(buffer)
+		c.request.Body = payload
+		c.request.ContentLength = int64(buffer.Len())
+		c.request.GetBody = func() (io.ReadCloser, error) {
+			if payload.Rewind() {
+				return payload, nil
+			}
+			return nil, errors.New("payload cannot be rewound")
+		}
+		// Wait for the payload to be fully drained before we return
+		// from Send to ensure the buffer can safely be reused.
+		defer payload.Wait()
```

Comment:
This seems excessively dangerous to have unbounded blocking here. Is it really 100% certain that the body will _always_ be drained or closed in a timely manner, in the face of all possible errors?

And, even if so, is it worth paying potential latency hit to wait for this? I would think instead we'd release the buffer (i.e. return to pool) asynchronously, instead of holding of the train here.

Reply:
`Close` has to be called as any reader would leak otherwise. This keeps the current behaviour between `sendUnary` and `sendStream`: both don't modify the buffer and both return and the buffer is safe to mutate afterwards.